ca_tools/scripts/control_node_proportional.py

```diff
@@ -0,0 +1,175 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2013 PAL Robotics SL.
+# Released under the BSD License.
+#
+# Authors:
+#   * Ramiro Serra
+
+import math
+import signal
+import sys
+from math import pi
+
+import rospy
+from geometry_msgs.msg import Twist
+from geometry_msgs.msg import Pose2D
+from geometry_msgs.msg import Point
+from nav_msgs.msg import Odometry
+from tf.transformations import euler_from_quaternion
+
+TURNING = 1
+STOP_TURNING = 2
+FORWARD = 3
+STOP_FORWARD = 4
+ANGULAR_VEL = 1.0
+LINEAR_VEL = 0.2
+KP = 2.5  # proportional coefficient
+BIG_ANGLE_THRESHOLD = 1.0
+DISTANCE_TOLERANCE = 0.2
+ANGLE_TOLERANCE = 0.2
+
+pubtopic = "/create1/cmd_vel"
+subtopic = "/create1/gts"
+
+
+def signal_handler(signal, frame, queue_size=1):
+  # This will execute when Ctrl-c is pressed
+    pub = rospy.Publisher(pubtopic, Twist, queue_size=1)
+    aux = Twist()
+    pub.publish(aux)
+    sys.exit(0)
+
+
+class Ctrl_Node:
+
+    def __init__(self):
+
+        rospy.init_node('control_node')
+
+        self._my_pub = rospy.Publisher(pubtopic, Twist, queue_size=10)
+
+        # Create a subscriber with appropriate topic, custom message and name of callback function.
+        self._my_sub = rospy.Subscriber(subtopic, Odometry, self._callback)
+        self._my_goals = []
+
+        for i in range(4):
+            self._my_goals.append(Pose2D())
+
+        self._my_goals[1].x = 2.0
+
+        self._my_goals[2].x = 2.0
+        self._my_goals[2].y = 2.0
+
+        self._my_goals[3].y = 2.0
+
+        self._goal_num = 0
+        self._state = STOP_FORWARD
+        self._my_pose = Pose2D()
+        self._my_pose.x = "unset"
+        self._my_pose.y = "unset"
+
+# Create a callback function for the subscriber.
+    def _callback(self, data):
+        self._my_pose.x = data.pose.pose.position.x
+        self._my_pose.y = data.pose.pose.position.y
+        q = data.pose.pose.orientation
+        q_arr = [q.x, q.y, q.z, q.w]
+        euler = euler_from_quaternion(q_arr)
+        self._my_pose.theta = euler[2]
+        self._move()
```

Comment:
This is not a good way of working with callbacks because they have to be as shorter as possible.
I recommend you to have a while loop in another method (called `run()`?) and based on the value of `_state`, make the robot move.

You'll also need to work with mutexes to make this better.